### PR TITLE
osd: flush/evict op fixes

### PR DESCRIPTION
**Pull request by wonzhq**
1) Move the skip promotion detection logic into init_op_flags
2) Skip promotion for flush/evict op
3) Correctly handle the evict op when obs doesn't exist

### CODE REVIEW

**Comment by liewegas** [comment]
This looks good to me.  One comment on the delete / FAILOK thing, though: the current version isn't quite correct because the return value for the delete op is still 0 when it should perhaps be ENOENT.  What we really need to do this properly is a new DELETE_IMMUTABLE op that returns 0 when the object doesn't exist, *or* we just change DELETE to do that.  We could only do the latter if we audited the code to make sure we don't have delete users who care about ENOENT, though, and made some noise about it in the release notes.

**Comment by liewegas** [comment]
Current version looks okay, though, modulo the whitespace; can you fix that and repush?

**Comment by wonzhq** [comment]
@liewegas updated. Also changed the test case accordingly.

File: src/osd/ReplicatedPG.cc [hunk 1]
```diff
@@ -1800,6 +1800,10 @@ bool ReplicatedPG::maybe_handle_cache(OpRequestRef op,
     osd->logger->inc(l_osd_op_cache_hit);
     return false;
   }
+  
+  if (op->need_skip_promote()) {
+    return false;
+  }
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: actually it is probably a bit safer to move this below the msut_promote case below.  I don't think they will conflict right now unless there is a notify + delete combined (whcih doesn't make sense), but we should avoid potential issues later.

Comment (wonzhq) [inline src/osd/ReplicatedPG.cc]: Yep, let's do this

**Comment by liewegas** [comment]
:+1:

**Pull request merged**